### ISSUE TITLE
Code generation for Postgres DBs assumes column data type to be in same schema as table

### ISSUE THREAD

**Issue by skroll** [jOOQ/jOOQ#4682]
I'm looking at extending some of jOOQ's code generation that depends on handling special cases when columns use the PostGIS types. I've noticed that iterating over a ColumnDefinition, the DataTypeDefinition returned by getType() returns the same schema as the table that it resides in, even if the type's schema is different. I see in PostgresTableDefinition, when creating the column type, it just calls getSchema() and passing it when creating a DefaultDataTypeDefinition.

It would be helpful to also get the ```UDT_SCHEMA``` column along with ```UDT_NAME``` when the data type is a UDT, and possibly setting that instead of just the table's schema.

I guess it would make sense if I showed what I was doing and expecting:
```
CREATE SCHEMA IF NOT EXISTS postgis;
CREATE EXTENSION IF NOT EXISTS postgis SCHEMA postgis;

CREATE TABLE public.zip_code (
  id SERIAL PRIMARY KEY,
  location postgis.geography(POINT, 4326)
);
```

Then during generation:
```
final SchemaDefinition schema = database.getSchema("public");
final TableDefinition table = database.getTable(schema, "zip_code");
final ColumnDefinition column = table.getColumn("location");
final DataTypeDefinition dataType = column.getType();
System.out.println(dataType.getSchema().toString());
```

I would expect the output to return "postgis", however it's returning "public", since it just assumes that the schema is the same as the table.

It would be nice if the UDT schema came along with the data type definition, since in it's current incarnation, I have to do some guess work to find out what the type is referencing.

I know this is probably a strange edge-case, but I'm trying to get our project to be able to be able to better handle some of the PostGIS functions automatically. Currently I'm using type converters/bindings to handle the geography/geometry type, however catching all the columns requires to explicitly list them in the codegen configuration, and it prevents us from using any of the functions that take a geography/geometry as a parameter. Currently I've jumped around this by implementing the functions as CustomFields, but it would be nice to be able to hook into the code generation and just check for <postgis_schema>.geometry to handle all of these in my generated code.

----
See also: #4707

**Comment by skroll** [comment]
For reference, this is where I'm seeing the schema being set:
https://github.com/jOOQ/jOOQ/blob/207528f704265577e0625e6f75b13949b15c7c7c/jOOQ-meta/src/main/java/org/jooq/util/postgres/PostgresTableDefinition.java#L103

**Comment by lukaseder** [comment]
Thank you very much for your report. There had been fixes along these lines in the past - I believe when `enum` arrays were concerned. I'll investigate whether those fixes were incomplete in the context of composit types / UDTs.

**Comment by lukaseder** [comment]
This has probably been fixed implicitly via #4708 and #4712. Will double-check with `postgis` types

**Comment by lukaseder** [comment]
I'm sorry for the delay in this. I had to re-install PostgreSQL 9.5 and Postgis to check this.

I'm running this query here:

```sql
select n.nspname, t.*
from pg_type as t
join pg_namespace as n on t.typnamespace = n.oid
where t.typname like '%geo%'
```

And I can see that the `geography` type is in the `public` schema. Isn't that what's expected? Interestingly, you're allowed to prefix the type with the extension name, but the type still seems to be located in the `public` schema. There's no `postgis` schema on my installation. Does this match your observations?

**Comment by lukaseder** [comment]
Hmm, I'm not sure if I'm getting this right yet. It doesn't look like `geography` is really an ordinary composite type... Will do more research.

**Comment by skroll** [comment]
Depends on how you set up postgis. I prefer not to have all the postgis functions and types polluting the public schema, so my setup for my database is done like this:

```
CREATE SCHEMA postgis;
CREATE EXTENSION postgis WITH SCHEMA postgis;
```

This causes all the extra types and functions to be placed in the postgis schema instead.

**Comment by lukaseder** [comment]
OK, I see - I wasn't aware of that.

Anyway, the fixes in #4708 and #4712 should have resolved this problem. They're committed to 3.8.0-SNAPSHOT, and will be merged to 3.7.2. Please, reopen if you still experience issues.
